alacritty/src/input.rs

```diff
@@ -939,102 +990,67 @@ impl<T: EventListener, A: ActionContext<T>> Processor<T, A> {
-        let mouse_state = self.cursor_state();
-        self.ctx.window().set_mouse_cursor(mouse_state);
-    }
-
-    /// Reset mouse cursor based on modifier and terminal state.
-    #[inline]
-    pub fn reset_mouse_cursor(&mut self) {
-        let mouse_state = self.cursor_state();
-        self.ctx.window().set_mouse_cursor(mouse_state);
-    }
-
-    /// Process a received character.
-    pub fn received_char(&mut self, c: char) {
-        let suppress_chars = *self.ctx.suppress_chars();
-
-        // Don't insert chars when we have IME running.
-        if self.ctx.display().ime.preedit().is_some() {
+        // Key bindings suppress the character input.
+        if self.process_key_bindings(&key) {
             return;
         }
 
-        // Handle hint selection over anything else.
-        if self.ctx.display().hint_state.active() && !suppress_chars {
-            self.ctx.hint_input(c);
+        // TODO maybe we should just accept text?
+        if self.ctx.search_active() {
+            for ch in text.chars() {
+                self.ctx.search_input(ch);
+            }
+
             return;
         }
 
-        // Pass keys to search and ignore them during `suppress_chars`.
-        let search_active = self.ctx.search_active();
-        if suppress_chars || search_active || self.ctx.terminal().mode().contains(TermMode::VI) {
-            if search_active && !suppress_chars {
-                self.ctx.search_input(c);
-            }
-
+        // Vi mode on its own doesn't have any input, the search input was done before.
+        if self.ctx.terminal().mode().contains(TermMode::VI) || text.is_empty() {
             return;
         }
 
         self.ctx.on_terminal_input_start();
 
-        let utf8_len = c.len_utf8();
-        let mut bytes = vec![0; utf8_len];
-        c.encode_utf8(&mut bytes[..]);
-
         #[cfg(not(target_os = "macos"))]
         let alt_send_esc = true;
 
-        // Don't send ESC when `OptionAsAlt` is used. This doesn't handle
-        // `Only{Left,Right}` variants due to inability to distinguish them.
         #[cfg(target_os = "macos")]
-        let alt_send_esc = self.ctx.config().window.option_as_alt != OptionAsAlt::None;
+        let alt_send_esc = {
+            let option_as_alt = self.ctx.config().window.option_as_alt;
+            // Only send ESC for an `Alt` key.
+            option_as_alt == OptionAsAlt::Both
+                || (option_as_alt == OptionAsAlt::OnlyLeft
+                    && self.ctx.modifiers().lalt_state() == ModifiersKeyState::Pressed)
+                || (option_as_alt == OptionAsAlt::OnlyRight
+                    && self.ctx.modifiers().ralt_state() == ModifiersKeyState::Pressed)
+        };
 
-        if alt_send_esc
-            && *self.ctx.received_count() == 0
-            && self.ctx.modifiers().alt()
-            && utf8_len == 1
-        {
-            bytes.insert(0, b'\x1b');
+        let mut bytes = Vec::with_capacity(text.as_bytes().len() + 1);
```

Comment:
```suggestion
        let mut bytes = Vec::with_capacity(text.len() + 1);
```

This is identical, is it not?